### ISSUE TITLE
Remove the CLI and add a Parameter Browser instead

### ISSUE THREAD

**Issue by skaman82** [skaman82/cleanflight-configurator#59]
This requires exposing all CLI parameters to MSP.

Here the concept (will be updated shortly)
![cf0_advanced](https://cloud.githubusercontent.com/assets/11768024/11608092/5de4f0b2-9b60-11e5-9ea3-a70a8c426265.jpg)

**Comment by skaman82** [comment]
Starts to look very good. @NightHawk32  How do we get the default values to show up?

![bildschirmfoto 2016-01-07 um 12 21 19](https://cloud.githubusercontent.com/assets/11768024/12169509/31d8eabc-b539-11e5-9c37-85d9938252bf.png)

**Comment by NightHawk32** [comment]
At the moment the default aren't provided by the firmware, the only thing that could be done is storing them in the configurator, together with the description and name of a parameter.
The parameters are identified via group and param id, which is unique for each parameter.

I also had an idea about the number of used values, this can be determined for the data type of the parameter. So the information is already there and only has to be used.

**Comment by hydra** [comment]
Yes, the serial_cli.c contains a large array which can be extracted for other purposes.  One of the problems that needs solving is the 'default value' problem.  OSD's and GUI's would all benefit from being able to know the default value and reset a parameter to it.

**Comment by skaman82** [comment]
@NightHawk32 See https://github.com/cleanflight/cleanflight/pull/1049 This might be useful

**Comment by NightHawk32** [comment]
@hydra please also have a look at the current concept of digitalentity over here: https://github.com/digitalentity/cleanflight/tree/msp-parameter-protocoll. I did some small changes on that in my fork and used that for testing.
@skaman82 thanks, will have a look at that.

**Comment by digitalentity** [comment]
@NightHawk32, @hydra we can probably do both parameter group upgrade and my parameter protocol. 
The disadvantage of parameter group is that configurator has to be aware how to handle the group itself.
This is solved by parameter protocol approach that enables the configurator to access individual parameter of some generic data type.
We can extend the PG concept with list of individual parameters within the group and probably even have min/max/default values add well.

**Comment by wanderzell** [comment]
Hi guys, glad to see there have been some progress in the "parameter browser" development. The needed functionality for all of us who use CLI to tweak the system.

When I saw the initial sketches of the parameter sheet my first thought was - omg someone has been using the APM Planner too much. I seems bit too heavy with all this grouping and stuff, yet it lacks the most basic functions like:
- easy way to quickly determine if the value is different from the default, quick way of resetting it
- different type of fields need different type of editing. numbers are often easy to adjust in slider, for example
- import / export function (what many people use the CLI for today)

So I created a "Revision 2" for it.

![advanced configuration](https://cloud.githubusercontent.com/assets/4578152/12890503/982f0924-ce8c-11e5-8c61-2232f09c5c22.png)

1. It is just flat list of parameters, without fancy grouping. 
1. No labels, just parameter names as they are described in the docs. Keeps referring and finding them easy and straight-forward. We could add hint boxes at some moment later, or description fields, but initially, for sake of getting nice functional alternative to CLI I think it is really not needed.
1. Primary navigation element is the search box, which filters the parameter list as you type.
1. As there is no labels or groups it is easy for user to understand what to type into search box.
1. Each parameter has it's type which shows appropriate editor field for it: toggle fields, select boxes and number fields.
1. Range is not just static text, but fully functional slider component which gives user visual indication of current value relation to min/max, at the same time allowing to change the value visually.
1. Default is also not just a static text, but visual cue to show whether the current value is same as default or not. If value differs from default the default value is highlighted, also a tiny "reset" button is shown to allow user to reset current value by clicking it.

Please review and give me your thoughts. 

If the concept is ok I can take the JS/CSS development to my shoulders.

**Comment by hydra** [comment]
ok @wanderzell looks nice.  first though, we need to finish cleanflight/cleanflight#1049 before we can do the js FC communication side of things since we need to finalize the firmware side of things first. i think.

unless you can think of a way where we can do them both in parallel?

**Comment by skaman82** [comment]
@wanderzell This looks nice, I like It even more than mine first sketch  – I never used APM in my life though :) The only thing I don't like so much is the export/import function. There is already a Backup and Restore button in the Setup tab. Those buttons should cover the whole system config, so there is no need to put it in the parameter browser.

**Comment by skaman82** [comment]
Also putting parameters in groups can make findings things much more faster :)

**Comment by skaman82** [comment]
Some parameters have more than one variable (like mixer settings) You might want to figure this one out.

**Comment by wanderzell** [comment]
@hydra  I think we can start work on the UI already. Just real binding between the screen and FC can be added once the parameter group thing is there. 

@skaman82 I agree Import / Export would not be there in first run and it is probably better to keep this discussion separate from the parameters thing. 

The idea behind these buttons is more elaborate than just plain dump import, however. Was thinking to have there a visual importer where user could see the comparison of existing and importing parameters, like in table, additionally have ability to "tick" which parameters to import and which ones to skip. That would follow more closely the way how people import CLI dumps these days. 

We can add grouped view as well, but lets start from the flat list as it is more straight-forward to implement and needed anyway (for filter).

**Comment by gaelj** [comment]
Any progress ?

**Comment by skaman82** [comment]
@gaelj  Sadly, no one is working on this at the moment - as far I know. I made a interface suggestion for this and there is initial work done by @NightHawk32. @wanderzell had a better UI suggestion, but I doubt that he has made some further progress on this.

If you want, you can try to get this done. It would be AWSOME

Some useful sources: 
https://github.com/NightHawk32/cleanflight/tree/msp-parameter-protocol
https://github.com/NightHawk32/cleanflight-configurator/tree/parameter_browser

**Comment by skaman82** [comment]
@gaelj since you found my CF issue tracker - or more a "things that would be nice to have in CF" take a look around, maybe you can crack a thing or two. :)

**Comment by wanderzell** [comment]
Actually I have done some work on this already. The UI and CSS is there, also the interaction related JS. But yeah, it has been standing in this state for few months now. I will try to pull myself together and merge the stuff to latest development in coming days.

**Comment by gaelj** [comment]
Ok ! Let me know if I can be of any assistance.

**Comment by skaman82** [comment]
@wanderzell some updates?

**Comment by wanderzell** [comment]
Yes there have been progress. Just give me more few more days to wrap up the initial UI implementation.

**Comment by DaleKramer** [comment]
Are you handling different profiles and rateprofiles?  I think Boris is down to 2 profiles and 3 rate profiles.

**Comment by wanderzell** [comment]
@DaleKramer : haven’t thought about profiles. But regarding rateprofiles I have two solutions in my mind:

1) in first run I’d simply show rateprofile related params based on currently selected rateprofile. so user must make sure the correct rateprofile is selected when making adjustments. I’d just add small visual notation to differentiate the parameter scope (global vs. profile)

2) in longer run I’d add a simple filter next to search box to narrow the parameter list by scope. Show: All / Global only / Rateprofile only. And make it also possible to show all profiles together so that parameters could be easily tweaked across the profiles without jumping between the views (this feature could be enabled or available only when rateprofile parameters are shown and/or user is filtering the view by parameter name. we don't want the parameter browser become too crowded for people who just use one single profile)

**Comment by skaman82** [comment]
@wanderzell Agree on 1 and 2

**Comment by DaleKramer** [comment]
Good idea on showing all 3 rateprofiles at once.  Nice if you could do the same with profiles.
Also I don't see anything about how to deal with parameters that have multiple values.
I think these changes should deal with all possibilities of handling CLI dumps from all forks and even then there are a lot of reasons for still letting CLI live in some form.

**Comment by wanderzell** [comment]
@DaleKramer : yep, personally I don’t have anything against CLI or the idea of keeping it. I use it and I know many people use it. Not because we like it that much, there is just no alternative. 

Parameter Browser is an attempt to provide a solid, cross-firmware compatible alternative for many things that CLI is better not used - retrieving and setting simple parameter values, for example. If it works then it may happen that we can keep CLI focused for rest of the stuff. May happen we figure out better UI for that other stuff as well so CLI becomes obsolete, who knows.

**Comment by DaleKramer** [comment]
Great, I'm all for making things easier, stages or not :)
